int_mm microbenchmark experiments

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111631
* #111413

Summary: 

---------------------------------------WORST---------------------------------------

test total time
| shape | torch int_mm | triton matmul | inductor aten int_mm | inductor triton int_mm | inductor triton bp int_mm |
|--------------------------|------|------|------|------|------|
|[78400, 3840] x [3840, 1280]|2.8103|1.6544|2.8128|1.7828|1.7566|
|[78400, 1280] x [1280, 1280]|1.0351|0.6128|1.0360|0.7081|0.6888|
|[65536, 5120] x [5120, 1280]|3.1008|1.8279|3.1213|1.9045|1.9116|
|[65536, 1280] x [1280, 5120]|2.6623|2.0272|2.6378|2.4017|2.4073|

test GPU time
| shape | torch int_mm | triton matmul | inductor aten int_mm | inductor triton int_mm | inductor triton bp int_mm |
|--------------------------|------|------|------|------|------|
|[78400, 3840] x [3840, 1280]|2.8413|1.6765|2.8672|1.7841|1.7865|
|[78400, 1280] x [1280, 1280]|1.0611|0.6202|1.0442|0.7261|0.7172|
|[65536, 5120] x [5120, 1280]|3.1531|1.8451|3.1657|1.9496|1.9466|
|[65536, 1280] x [1280, 5120]|2.7056|2.0416|2.6854|2.3970|2.4349|

---------------------------------------NEW CONFIGS, NO BP BRANCH, GROUP_M 4, NO COORD_DESC------------------

test total time
| shape | torch int_mm | triton matmul | inductor aten int_mm | inductor triton int_mm | inductor triton bp int_mm |
|--------------------------|------|------|------|------|------|
|[78400, 3840] x [3840, 1280]|2.8245|1.6619|2.8141|1.7536|1.7524|
|[78400, 1280] x [1280, 1280]|1.0218|0.6109|1.0341|0.6977|0.6929|
|[65536, 5120] x [5120, 1280]|3.1039|1.8319|3.1220|1.8981|1.8963|
|[65536, 1280] x [1280, 5120]|2.6539|2.0095|2.6515|2.3723|2.3612|

test GPU time
| shape | torch int_mm | triton matmul | inductor aten int_mm | inductor triton int_mm | inductor triton bp int_mm |
|--------------------------|------|------|------|------|------|
|[78400, 3840] x [3840, 1280]|2.7450|1.6660|2.8623|1.7786|1.7829|
|[78400, 1280] x [1280, 1280]|1.0656|0.6223|1.0510|0.7029|0.7046|
|[65536, 5120] x [5120, 1280]|3.1542|1.8781|3.1473|1.9223|1.9042|
|[65536, 1280] x [1280, 5120]|2.6847|2.1089|2.6759|2.4057|2.3827|

---------------------------------------BEST--------------------------------------

test total time
| shape | torch int_mm | triton matmul | inductor aten int_mm | inductor triton int_mm | inductor triton bp int_mm |
|--------------------------|------|------|------|------|------|
|[78400, 3840] x [3840, 1280]|2.8240|1.6727|2.8278|1.7537|1.7619|
|[78400, 1280] x [1280, 1280]|1.0632|0.6032|1.0367|0.6984|0.6945|
|[65536, 5120] x [5120, 1280]|3.1099|1.8358|3.0890|1.9131|1.9114|
|[65536, 1280] x [1280, 5120]|2.6537|2.0301|2.6527|2.3573|2.3718|

test GPU time
| shape | torch int_mm | triton matmul | inductor aten int_mm | inductor triton int_mm | inductor triton bp int_mm |
|--------------------------|------|------|------|------|------|
|[78400, 3840] x [3840, 1280]|2.8312|1.6641|2.8678|1.7451|1.7754|
|[78400, 1280] x [1280, 1280]|1.0573|0.6212|1.0472|0.7091|0.6916|
|[65536, 5120] x [5120, 1280]|3.1163|1.8380|3.1533|1.9126|1.9305|
|[65536, 1280] x [1280, 5120]|2.6471|2.0384|2.6961|2.3507|2.3852|

-----------------------------------JUST COORD DESC------------------------------

test total time
| shape | torch int_mm | triton matmul | inductor aten int_mm | inductor triton int_mm | inductor triton bp int_mm |
|--------------------------|------|------|------|------|------|
|[78400, 3840] x [3840, 1280]|2.8202|1.6694|2.8441|1.7868|1.7732|
|[78400, 1280] x [1280, 1280]|1.0209|0.6205|1.0397|0.7124|0.7123|
|[65536, 5120] x [5120, 1280]|3.1385|1.8373|3.1357|1.9221|1.9253|
|[65536, 1280] x [1280, 5120]|2.6637|2.0442|2.6747|2.4135|2.4333|

test GPU time
| shape | torch int_mm | triton matmul | inductor aten int_mm | inductor triton int_mm | inductor triton bp int_mm |
|--------------------------|------|------|------|------|------|
|[78400, 3840] x [3840, 1280]|2.8710|1.6757|2.8686|1.7988|1.7988|
|[78400, 1280] x [1280, 1280]|1.0599|0.6262|1.0436|0.7143|0.7246|
|[65536, 5120] x [5120, 1280]|3.2012|1.8360|3.1548|1.9311|1.9487|
|[65536, 1280] x [1280, 5120]|2.6890|2.0702|2.6873|2.4376|2.4373|

(78400, 1280, 3840) BLOCK_M: 128, BLOCK_N: 256, BLOCK_K: 128, SPLIT_K: 1, num_warps: 8, num_ctas: 1, num_stages: 3, enable_warp_specialization: False, enable_persistent: False
(78400, 1280, 1280) BLOCK_M: 128, BLOCK_N: 256, BLOCK_K: 128, SPLIT_K: 1, num_warps: 8, num_ctas: 1, num_stages: 3, enable_warp_specialization: False, enable_persistent: False
(65536, 1280, 5120) BLOCK_M: 128, BLOCK_N: 256, BLOCK_K: 128, SPLIT_K: 1, num_warps: 8, num_ctas: 1, num_stages: 3, enable_warp_specialization: False, enable_persistent: False
(65536, 5120, 1280) BLOCK_M: 128, BLOCK_N: 256, BLOCK_K: 128, SPLIT_K: 1, num_warps: 8, num_ctas: 1, num_stages: 3, enable_warp_specialization: False, enable_persistent: False

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

HDCharles committed:
int_mm microbenchmark experiments

Summary: block pointers, new configs and group_m=4 possible

test total time
shape; torch int_mm; triton matmul; inductor aten int_mm; inductor triton int_mm; inductor triton bp int_mm
[128, 9216] x [9216, 4096]; 0.058004120364785194; 0.10535435751080513; 0.073293661698699; 0.07340429350733757; 0.07662827149033546
[128, 4096] x [4096, 4096]; 0.024163806810975075; 0.10901077650487423; 0.07408992387354374; 0.0722301471978426; 0.07132194936275482
[128, 4096] x [4096, 1000]; 0.036254385486245155; 0.10616344399750233; 0.07640384137630463; 0.07024327293038368; 0.0717272236943245
[2048, 768] x [768, 768]; 0.018268311396241188; 0.10573056526482105; 0.07481606677174568; 0.07651427760720253; 0.07615952752530575
[2048, 768] x [768, 3072]; 0.04023437388241291; 0.11132276616990566; 0.07324442267417908; 0.07273463532328606; 0.06787577643990517
[2048, 3072] x [3072, 768]; 0.043066078796982765; 0.10560395196080208; 0.07386437617242336; 0.07078452967107296; 0.06927150301635265
[1024, 768] x [768, 768]; 0.015496015548706056; 0.10603046976029873; 0.07354759611189365; 0.0758734904229641; 0.07564103230834007
[1024, 768] x [768, 3072]; 0.022138087078928947; 0.10170646943151951; 0.06307680159807205; 0.06898500956594944; 0.06927525624632835
[1024, 3072] x [3072, 768]; 0.028426432982087135; 0.09982505813241005; 0.07412933744490147; 0.068456856533885; 0.06740028038620949
[1024, 768] x [768, 2304]; 0.021387580782175064; 0.1016910932958126; 0.07339306175708771; 0.0769086554646492; 0.07137617096304893
test GPU time
shape; torch int_mm; triton matmul; inductor aten int_mm; inductor triton int_mm; inductor triton bp int_mm
[128, 9216] x [9216, 4096]; 0.06933990865945816; 0.06610661745071411; 0.06979862600564957; 0.07433804869651794; 0.07468699663877487
[128, 4096] x [4096, 4096]; 0.03411557525396347; 0.033387377858161926; 0.03366631641983986; 0.03769776597619057; 0.037724412977695465
[128, 4096] x [4096, 1000]; 0.04040428251028061; 0.025955677032470703; 0.04039134085178375; 0.03547864779829979; 0.03545495495200157
[2048, 768] x [768, 768]; 0.02109251357614994; 0.014950398355722427; 0.02105111815035343; 0.01599065214395523; 0.016331763938069344
[2048, 768] x [768, 3072]; 0.044213179498910904; 0.0345146544277668; 0.04417350888252258; 0.040452226996421814; 0.04010145738720894
[2048, 3072] x [3072, 768]; 0.04449857398867607; 0.03184685856103897; 0.04453713446855545; 0.03114512376487255; 0.031160898506641388
[1024, 768] x [768, 768]; 0.015048310160636902; 0.01198173500597477; 0.015472636558115482; 0.014197333715856075; 0.014555573463439941
[1024, 768] x [768, 3072]; 0.02595936693251133; 0.020467696711421013; 0.02595866098999977; 0.02351984567940235; 0.023510659113526344
[1024, 3072] x [3072, 768]; 0.032588981091976166; 0.02414623275399208; 0.03259038180112839; 0.029331060126423836; 0.02935468591749668
[1024, 768] x [768, 2304]; 0.025275489315390587; 0.019762994721531868; 0.025669628754258156; 0.021082209423184395; 0.021062878891825676

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "int_mm microbenchmark experiments"


Summary: block pointers, new configs and group_m=4 possible

---------------------------------------WITHOUT GROUP_M=4---------------------------------------
test total time
|shape| torch int_mm| triton matmul| inductor aten int_mm| inductor triton int_mm| inductor triton bp int_mm|
|-----|-------------|--------------|---------------------|-----------------------|--------------------------|
|[128, 9216] x [9216, 4096]| 0.05799| 0.10517| 0.06817| 0.06909| 0.07094|
|[128, 4096] x [4096, 4096]| 0.02416| 0.11976| 0.07393| 0.07287| 0.07042|
|[128, 4096] x [4096, 1000]| 0.03626| 0.11042| 0.06695| 0.07042| 0.07709|
|[2048, 768] x [768, 768]| 0.01828| 0.10658| 0.06808| 0.07399| 0.07611|
|[2048, 768] x [768, 3072]| 0.04024| 0.10827| 0.06870| 0.07038| 0.07122|
|[2048, 3072] x [3072, 768]| 0.04303| 0.10412| 0.06950| 0.07113| 0.06890|
|[1024, 768] x [768, 768]| 0.01534| 0.10368| 0.06642| 0.06732| 0.07504|
|[1024, 768] x [768, 3072]| 0.02212| 0.10721| 0.06806| 0.07025| 0.07209|
|[1024, 3072] x [3072, 768]| 0.02843| 0.10040| 0.07356| 0.07597| 0.06497|
|[1024, 768] x [768, 2304]| 0.02142| 0.10798| 0.07265| 0.07583| 0.07577|
test GPU time
shape| torch int_mm| triton matmul| inductor aten int_mm| inductor triton int_mm| inductor triton bp int_mm
|-----|-------------|--------------|---------------------|-----------------------|--------------------------|
|[128, 9216] x [9216, 4096]| 0.06937| 0.06606| 0.06941| 0.07494| 0.07445|
|[128, 4096] x [4096, 4096]| 0.03367| 0.03306| 0.03361| 0.03770| 0.03811|
|[128, 4096] x [4096, 1000]| 0.04053| 0.02642| 0.04048| 0.03552| 0.03511|
|[2048, 768] x [768, 768]| 0.02109| 0.01495| 0.02114| 0.01635| 0.01636|
|[2048, 768] x [768, 3072]| 0.04425| 0.03453| 0.04477| 0.04079| 0.04039|
|[2048, 3072] x [3072, 768]| 0.04436| 0.03091| 0.04394| 0.03152| 0.03151|
|[1024, 768] x [768, 768]| 0.01504| 0.01204| 0.01547| 0.01421| 0.01459|
|[1024, 768] x [768, 3072]| 0.02599| 0.02061| 0.02597| 0.02402| 0.02361|
|[1024, 3072] x [3072, 768]| 0.03218| 0.02463| 0.03252| 0.02938| 0.02937|
|[1024, 768] x [768, 2304]| 0.02563| 0.01982| 0.02522| 0.02072| 0.02074|

---------------------------------------ALLOW GROUP_M=4---------------------------------------
test total time
|shape| torch int_mm| triton matmul| inductor aten int_mm| inductor triton int_mm| inductor triton bp int_mm|
|-----|-------------|--------------|---------------------|-----------------------|--------------------------|
|[128, 9216] x [9216, 4096]| 0.05800| 0.10535| 0.07329| 0.07340| 0.07662|
|[128, 4096] x [4096, 4096]| 0.02416| 0.10901| 0.07408| 0.07223| 0.07132|
|[128, 4096] x [4096, 1000]| 0.03625| 0.10616| 0.07640| 0.07024| 0.07172|
|[2048, 768] x [768, 768]| 0.01826| 0.10573| 0.07481| 0.07651| 0.07615|
|[2048, 768] x [768, 3072]| 0.04023| 0.11132| 0.07324| 0.07273| 0.06787|
|[2048, 3072] x [3072, 768]| 0.04306| 0.10560| 0.07386| 0.07078| 0.06927|
|[1024, 768] x [768, 768]| 0.01549| 0.10603| 0.07354| 0.07587| 0.07564|
|[1024, 768] x [768, 3072]| 0.02213| 0.10170| 0.06307| 0.06898| 0.06927|
|[1024, 3072] x [3072, 768]| 0.02842| 0.09982| 0.07412| 0.06845| 0.06740|
|[1024, 768] x [768, 2304]| 0.02138| 0.10169| 0.07339| 0.07690| 0.07137|
test GPU time
|shape| torch int_mm| triton matmul| inductor aten int_mm| inductor triton int_mm| inductor triton bp int_mm|
|-----|-------------|--------------|---------------------|-----------------------|--------------------------|
|[128, 9216] x [9216, 4096]| 0.06933| 0.06610| 0.06979| 0.07433| 0.07468|
|[128, 4096] x [4096, 4096]| 0.03411| 0.03338| 0.03366| 0.03769| 0.03772|
|[128, 4096] x [4096, 1000]| 0.04040| 0.02595| 0.04039| 0.03547| 0.03545|
|[2048, 768] x [768, 768]| 0.02109| 0.01495| 0.02105| 0.01599| 0.01633|
|[2048, 768] x [768, 3072]| 0.04421| 0.03451| 0.04417| 0.04045| 0.04010|
|[2048, 3072] x [3072, 768]| 0.04449| 0.03184| 0.04453| 0.03114| 0.03116|
|[1024, 768] x [768, 768]| 0.01504| 0.01198| 0.01547| 0.01419| 0.01455|
|[1024, 768] x [768, 3072]| 0.02595| 0.02046| 0.02595| 0.02351| 0.02351|
|[1024, 3072] x [3072, 768]| 0.03258| 0.02414| 0.03259| 0.02933| 0.02935|
|[1024, 768] x [768, 2304]| 0.02527| 0.01976| 0.02566| 0.02108| 0.02106|

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

HDCharles committed:
Update on "int_mm microbenchmark experiments"


Summary: block pointers, new configs and group_m=4 possible

---------------------------------------ALLOW GROUP_M=4---------------------------------------
test GPU time
| shape | torch int_mm | triton matmul | inductor aten int_mm | inductor triton int_mm | inductor triton bp int_mm |
|--------------------------|------|------|------|------|------|
|[128, 9216] x [9216, 4096]|0.0692|0.0653|0.0696|0.0743|0.0743|
|[128, 4096] x [4096, 4096]|0.0335|0.0331|0.0339|0.0377|0.0381|
|[128, 4096] x [4096, 1000]|0.0406|0.0261|0.0406|0.0356|0.0352|
|[2048, 768] x [768, 768]|0.0210|0.0150|0.0210|0.0161|0.0166|
|[2048, 768] x [768, 3072]|0.0448|0.0341|0.0449|0.0400|0.0404|
|[2048, 3072] x [3072, 768]|0.0450|0.0315|0.0446|0.0317|0.0317|
|[1024, 768] x [768, 768]|0.0154|0.0116|0.0151|0.0146|0.0142|
|[1024, 768] x [768, 3072]|0.0264|0.0206|0.0264|0.0239|0.0239|
|[1024, 3072] x [3072, 768]|0.0322|0.0242|0.0324|0.0290|0.0294|
|[1024, 768] x [768, 2304]|0.0255|0.0199|0.0252|0.0212|0.0208|

---------------------------------------WITHOUT GROUP_M=4---------------------------------------
test GPU time
| shape | torch int_mm | triton matmul | inductor aten int_mm | inductor triton int_mm | inductor triton bp int_mm |
|--------------------------|------|------|------|------|------|
|[128, 9216] x [9216, 4096]|0.0694|0.0661|0.0693|0.0744|0.0744|
|[128, 4096] x [4096, 4096]|0.0336|0.0334|0.0337|0.0378|0.0378|
|[128, 4096] x [4096, 1000]|0.0409|0.0261|0.0409|0.0356|0.0352|
|[2048, 768] x [768, 768]|0.0211|0.0150|0.0210|0.0164|0.0160|
|[2048, 768] x [768, 3072]|0.0443|0.0341|0.0448|0.0405|0.0405|
|[2048, 3072] x [3072, 768]|0.0441|0.0310|0.0440|0.0316|0.0312|
|[1024, 768] x [768, 768]|0.0151|0.0118|0.0155|0.0147|0.0143|
|[1024, 768] x [768, 3072]|0.0264|0.0206|0.0264|0.0236|0.0237|
|[1024, 3072] x [3072, 768]|0.0326|0.0242|0.0322|0.0292|0.0292|
|[1024, 768] x [768, 2304]|0.0250|0.0198|0.0250|0.0208|0.0212|

---------------------------------------WITHOUT GROUP_M=4 and without bp kernel branches---------------------------------------
test GPU time
| shape | torch int_mm | triton matmul | inductor aten int_mm | inductor triton int_mm | inductor triton bp int_mm |
|--------------------------|------|------|------|------|------|
|[128, 9216] x [9216, 4096]|0.0694|0.0664|0.0694|0.0748|0.0747|
|[128, 4096] x [4096, 4096]|0.0335|0.0330|0.0337|0.0381|0.0377|
|[128, 4096] x [4096, 1000]|0.0405|0.0264|0.0408|0.0355|0.0355|
|[2048, 768] x [768, 768]|0.0211|0.0153|0.0216|0.0164|0.0160|
|[2048, 768] x [768, 3072]|0.0441|0.0341|0.0441|0.0403|0.0403|
|[2048, 3072] x [3072, 768]|0.0446|0.0319|0.0445|0.0312|0.0316|
|[1024, 768] x [768, 768]|0.0154|0.0152|0.0154|0.0147|0.0147|
|[1024, 768] x [768, 3072]|0.0261|0.0203|0.0265|0.0240|0.0236|
|[1024, 3072] x [3072, 768]|0.0323|0.0245|0.0323|0.0296|0.0296|
|[1024, 768] x [768, 2304]|0.0255|0.0198|0.0255|0.0212|0.0213|


---------------------------------------WITHOUT GROUP_M=4 and without new configs---------------------------------------

test GPU time
| shape | torch int_mm | triton matmul | inductor aten int_mm | inductor triton int_mm | inductor triton bp int_mm |
|--------------------------|------|------|------|------|------|
|[128, 9216] x [9216, 4096]|0.0693|0.0660|0.0697|0.0743|0.0743|
|[128, 4096] x [4096, 4096]|0.0341|0.0335|0.0340|0.0381|0.0377|
|[128, 4096] x [4096, 1000]|0.0408|0.0260|0.0404|0.0353|0.0352|
|[2048, 768] x [768, 768]|0.0210|0.0154|0.0210|0.0212|0.0208|
|[2048, 768] x [768, 3072]|0.0443|0.0345|0.0443|0.0452|0.0448|
|[2048, 3072] x [3072, 768]|0.0441|0.0313|0.0446|0.0499|0.0503|
|[1024, 768] x [768, 768]|0.0154|0.0117|0.0150|0.0142|0.0142|
|[1024, 768] x [768, 3072]|0.0265|0.0205|0.0261|0.0263|0.0263|
|[1024, 3072] x [3072, 768]|0.0321|0.0246|0.0321|0.0325|0.0321|
|[1024, 768] x [768, 2304]|0.0250|0.0195|0.0250|0.0238|0.0239|

---------------------------------------ALLOW GROUP_M=4 but without new configs---------------------------------------

test GPU time
| shape | torch int_mm | triton matmul | inductor aten int_mm | inductor triton int_mm | inductor triton bp int_mm |
|--------------------------|------|------|------|------|------|
|[128, 9216] x [9216, 4096]|0.0696|0.0666|0.0697|0.0747|0.0747|
|[128, 4096] x [4096, 4096]|0.0337|0.0331|0.0341|0.0381|0.0377|
|[128, 4096] x [4096, 1000]|0.0409|0.0264|0.0408|0.0356|0.0356|
|[2048, 768] x [768, 768]|0.0211|0.0154|0.0210|0.0214|0.0210|
|[2048, 768] x [768, 3072]|0.0445|0.0345|0.0447|0.0450|0.0446|
|[2048, 3072] x [3072, 768]|0.0442|0.0310|0.0442|0.0500|0.0499|
|[1024, 768] x [768, 768]|0.0154|0.0120|0.0154|0.0142|0.0146|
|[1024, 768] x [768, 3072]|0.0265|0.0202|0.0265|0.0256|0.0256|
|[1024, 3072] x [3072, 768]|0.0321|0.0244|0.0321|0.0325|0.0321|
|[1024, 768] x [768, 2304]|0.0254|0.0200|0.0250|0.0234|0.0238|
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

cpuhrsch commented on benchmarks/dynamo/microbenchmarks/inductor_int_mm.py:
```diff
@@ -0,0 +1,165 @@
+        b = torch.randint(-128, 127, b_shape, device="cuda", dtype=a.dtype).t().contiguous().t()
+
+        torch_ms = triton.testing.do_bench(lambda: torch_mm(a, b))
+        triton_ms = triton.testing.do_bench(lambda: triton_mm(a, b))
```
Seems like this one is the fastest?

cpuhrsch replied:
Maybe they use a lot more configs https://github.com/openai/triton/blob/main/python/triton/ops/matmul.py#L46-L67 ?

HDCharles committed:
Update on "int_mm microbenchmark experiments"


Summary: block pointers, new configs and group_m=4 possible

---------------------------------------ALLOW GROUP_M=4---------------------------------------
test GPU time
| shape | torch int_mm | triton matmul | inductor aten int_mm | inductor triton int_mm | inductor triton bp int_mm |
|--------------------------|------|------|------|------|------|
|[128, 9216] x [9216, 4096]|0.0692|0.0653|0.0696|0.0743|0.0743|
|[128, 4096] x [4096, 4096]|0.0335|0.0331|0.0339|0.0377|0.0381|
|[128, 4096] x [4096, 1000]|0.0406|0.0261|0.0406|0.0356|0.0352|
|[2048, 768] x [768, 768]|0.0210|0.0150|0.0210|0.0161|0.0166|
|[2048, 768] x [768, 3072]|0.0448|0.0341|0.0449|0.0400|0.0404|
|[2048, 3072] x [3072, 768]|0.0450|0.0315|0.0446|0.0317|0.0317|
|[1024, 768] x [768, 768]|0.0154|0.0116|0.0151|0.0146|0.0142|
|[1024, 768] x [768, 3072]|0.0264|0.0206|0.0264|0.0239|0.0239|
|[1024, 3072] x [3072, 768]|0.0322|0.0242|0.0324|0.0290|0.0294|
|[1024, 768] x [768, 2304]|0.0255|0.0199|0.0252|0.0212|0.0208|

---------------------------------------WITHOUT GROUP_M=4---------------------------------------
test GPU time
| shape | torch int_mm | triton matmul | inductor aten int_mm | inductor triton int_mm | inductor triton bp int_mm |
|--------------------------|------|------|------|------|------|
|[128, 9216] x [9216, 4096]|0.0694|0.0661|0.0693|0.0744|0.0744|
|[128, 4096] x [4096, 4096]|0.0336|0.0334|0.0337|0.0378|0.0378|
|[128, 4096] x [4096, 1000]|0.0409|0.0261|0.0409|0.0356|0.0352|
|[2048, 768] x [768, 768]|0.0211|0.0150|0.0210|0.0164|0.0160|
|[2048, 768] x [768, 3072]|0.0443|0.0341|0.0448|0.0405|0.0405|
|[2048, 3072] x [3072, 768]|0.0441|0.0310|0.0440|0.0316|0.0312|
|[1024, 768] x [768, 768]|0.0151|0.0118|0.0155|0.0147|0.0143|
|[1024, 768] x [768, 3072]|0.0264|0.0206|0.0264|0.0236|0.0237|
|[1024, 3072] x [3072, 768]|0.0326|0.0242|0.0322|0.0292|0.0292|
|[1024, 768] x [768, 2304]|0.0250|0.0198|0.0250|0.0208|0.0212|

---------------------------------------WITHOUT GROUP_M=4 and without bp kernel branches---------------------------------------
test GPU time
| shape | torch int_mm | triton matmul | inductor aten int_mm | inductor triton int_mm | inductor triton bp int_mm |
|--------------------------|------|------|------|------|------|
|[128, 9216] x [9216, 4096]|0.0694|0.0664|0.0694|0.0748|0.0747|
|[128, 4096] x [4096, 4096]|0.0335|0.0330|0.0337|0.0381|0.0377|
|[128, 4096] x [4096, 1000]|0.0405|0.0264|0.0408|0.0355|0.0355|
|[2048, 768] x [768, 768]|0.0211|0.0153|0.0216|0.0164|0.0160|
|[2048, 768] x [768, 3072]|0.0441|0.0341|0.0441|0.0403|0.0403|
|[2048, 3072] x [3072, 768]|0.0446|0.0319|0.0445|0.0312|0.0316|
|[1024, 768] x [768, 768]|0.0154|0.0152|0.0154|0.0147|0.0147|
|[1024, 768] x [768, 3072]|0.0261|0.0203|0.0265|0.0240|0.0236|
|[1024, 3072] x [3072, 768]|0.0323|0.0245|0.0323|0.0296|0.0296|
|[1024, 768] x [768, 2304]|0.0255|0.0198|0.0255|0.0212|0.0213|


---------------------------------------WITHOUT GROUP_M=4 and without new configs---------------------------------------

test GPU time
| shape | torch int_mm | triton matmul | inductor aten int_mm | inductor triton int_mm | inductor triton bp int_mm |
|--------------------------|------|------|------|------|------|
|[128, 9216] x [9216, 4096]|0.0693|0.0660|0.0697|0.0743|0.0743|
|[128, 4096] x [4096, 4096]|0.0341|0.0335|0.0340|0.0381|0.0377|
|[128, 4096] x [4096, 1000]|0.0408|0.0260|0.0404|0.0353|0.0352|
|[2048, 768] x [768, 768]|0.0210|0.0154|0.0210|0.0212|0.0208|
|[2048, 768] x [768, 3072]|0.0443|0.0345|0.0443|0.0452|0.0448|
|[2048, 3072] x [3072, 768]|0.0441|0.0313|0.0446|0.0499|0.0503|
|[1024, 768] x [768, 768]|0.0154|0.0117|0.0150|0.0142|0.0142|
|[1024, 768] x [768, 3072]|0.0265|0.0205|0.0261|0.0263|0.0263|
|[1024, 3072] x [3072, 768]|0.0321|0.0246|0.0321|0.0325|0.0321|
|[1024, 768] x [768, 2304]|0.0250|0.0195|0.0250|0.0238|0.0239|

---------------------------------------ALLOW GROUP_M=4 but without new configs---------------------------------------

test GPU time
| shape | torch int_mm | triton matmul | inductor aten int_mm | inductor triton int_mm | inductor triton bp int_mm |
|--------------------------|------|------|------|------|------|
|[128, 9216] x [9216, 4096]|0.0696|0.0666|0.0697|0.0747|0.0747|
|[128, 4096] x [4096, 4096]|0.0337|0.0331|0.0341|0.0381|0.0377|
|[128, 4096] x [4096, 1000]|0.0409|0.0264|0.0408|0.0356|0.0356|
|[2048, 768] x [768, 768]|0.0211|0.0154|0.0210|0.0214|0.0210|
|[2048, 768] x [768, 3072]|0.0445|0.0345|0.0447|0.0450|0.0446|
|[2048, 3072] x [3072, 768]|0.0442|0.0310|0.0442|0.0500|0.0499|
|[1024, 768] x [768, 768]|0.0154|0.0120|0.0154|0.0142|0.0146|
|[1024, 768] x [768, 3072]|0.0265|0.0202|0.0265|0.0256|0.0256|
|[1024, 3072] x [3072, 768]|0.0321|0.0244|0.0321|0.0325|0.0321|
|[1024, 768] x [768, 2304]|0.0254|0.0200|0.0250|0.0234|0.0238|
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

cpuhrsch commented:
This PR does two things
1. Microbenchmarks
2. Add more configs to the templates

Can we split that into two so we can ablate the impact of 2 and then potentially land it?
